base/gmp.jl

```diff
@@ -41,11 +41,30 @@ type BigInt <: Integer
     size::Cint
     d::Ptr{Limb}
     function BigInt()
+        if length(bigintpool) > 0
+            b = pop!(bigintpool)
+            finalizer(b, poolingfinalizer)
+            #warn("\n\nREUSING!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n\n")
+            #@show b
+            return b
+        end
         b = new(zero(Cint), zero(Cint), C_NULL)
         ccall((:__gmpz_init,:libgmp), Void, (Ptr{BigInt},), &b)
-        finalizer(b, _gmp_clear_func)
+        finalizer(b, poolingfinalizer)
         return b
     end
+    BigInt(alloc, size, d) = new(alloc, size, d)
+end
+
+const bigintpool = BigInt[]
+const BIGINTPOOL_MAXSIZE = 100000
+function poolingfinalizer(b::BigInt)
+    if length(bigintpool) < BIGINTPOOL_MAXSIZE
+        push!(bigintpool, BigInt(b.alloc, b.size, b.d))
```

Comment:
A bit annoying that this requires allocating a new `BigInt` on the heap, a consequence of the annoying fact that the `BigInt` type is mutable (see also #9826).  Might be better to have a
```
immutable _immutable_BigInt
    alloc::Cint
    size::Cint
    d::Ptr{Limb}
end
```
type mirroring `BigInt`, and then make `bigintpool = _immutable_BigInt[]` instead.